minpy/primitive.py

```diff
@@ -10,25 +10,71 @@
 
 from . import array
 from .array_variants import ArrayType
+from .array_variants import variants
 from . import context
 from .utils import log
 from . import tape
 
-_logger = log.get_logger(__name__) # pylint: disable= invalid-name
+_logger = log.get_logger(__name__)  # pylint: disable= invalid-name
 GradFunc = collections.namedtuple('GradFunc', ['f', 'multi_grad_indices'])
 
+
+class NoGradientFuncError(ValueError):
+    """Error of no required gradient func defined for back propagation.
+
+    Parameters
+    ----------
+    func_name : str
+        Name of the primitive without gradient definition.
+    arg : str or int
+        If str, arg is the name of a keyword argument. If int, arg is the
+        position of a positional argument.
+    """
+
+    def __init__(self, func_name, arg):
+        if isinstance(arg, str):
+            arg_name = 'keyword argument'
+        else:
+            arg_name = 'argument'
+        msg = "No partial derivative of {} defined on {} {}.".format(func_name, arg_name, arg)
+        super(NoGradientFuncError, self).__init__(msg)
+
+
+class FakeGradFunc(object):
+    """Fake gradient func for raising NoGradientFuncError at proper place.
+
+    Parameters
+    ----------
+    func_name : str
+        Name of the primitive without gradient definition.
+    arg : str or int
+        If str, arg is the name of a keyword argument. If int, arg is the
+        position of a positional argument.
+    """
+
+    def __init__(self, func_name, arg):
+        self.func_name = func_name
+        self.arg = arg
+
+    def __call__(self, grad_value):
+        raise NoGradientFuncError(self.func_name, self.arg)
+
+
 class Primitive(object):
     """Class for primitives. It includes both forward function and gradient definition."""
+    # pylint: disable=too-many-arguments
     __slots__ = [
         '_func',
         '_grad_func',
         '_grad_func_kw',
         '_type',
         '_mutate_args',
         '_mutate_kw',
+        '_custom_op',
+        '__name__',
     ]
 
-    def __init__(self, func, ty, mutate_args=None, mutate_kw=None):
+    def __init__(self, func, ty, mutate_args=None, mutate_kw=None, custom_op=False):
```

Comment:
Ok, let's do this. We change this directly to `typestr`. The default is None, in which it will be inferred by the `ty`. Otherwise, we just directly use it for debugging.